src/macros.jl

```diff
@@ -237,7 +237,126 @@ macro grad(expr)
         end
     end |> esc
 end
+
+
+function funcall_fwd_to_rule(expr)
```

Comment:
Can you add some explanations and comments to the code?

Reply:
updated.